src/analysis/lattice.h

```diff
@@ -0,0 +1,62 @@
+#ifndef wasm_analysis_lattice_h
+#define wasm_analysis_lattice_h
+
+#include "wasm.h"
+#include <bitset>
+#include <iostream>
+
+namespace wasm::analysis {
+
+enum LatticeComparison { NO_RELATION, EQUAL, LESS, GREATER };
+
+template<typename T>
+constexpr bool has_compare = std::is_invocable_r<LatticeComparison,
+                                                 decltype(T::compare),
+                                                 const T&,
+                                                 const T&>::value;
+template<typename T>
+constexpr bool has_getLeastUpperBound = std::
+  is_invocable_r<bool, decltype(&T::getLeastUpperBound), T, const T&>::value;
```

Comment:
Since this method returns `void`, I wouldn't expect this assertion to supply a return type.

```suggestion
constexpr bool has_getLeastUpperBound = std::
  is_invocable<decltype(&T::getLeastUpperBound), T, const T&>::value;
```

Reply:
This is a bit of a mix-up. I planned to convert the function to return a bool in later changes because this would let me know whether the lattice element was changed by the LUB, which is convenient when determining which nodes to queue in the worklist. However, I will change it back here.

Reply:
Makes sense 👍

Reply:
I've removed the static `getLeastUpperBound` function because the static assert constexpr doesn't seem to like the fact that its name is overloaded with the `getLeastUpperBound` member function and throws compile errors.